crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1854,21 +1875,87 @@ impl<'db> ClassType<'db> {
     }
 
+    /// Iterate over all superclasses of this class, excluding this class itself.
+    fn proper_superclasses(self, db: &'db dyn Db) -> impl Iterator<Item = ClassBase<'db>> {
+        self.mro(db).elements().skip(1)
+    }
+
     pub(crate) fn inherited_class_member(self, db: &'db dyn Db, name: &str) -> Symbol<'db> {
-        for base in self.bases(db) {
-            let member = base.member(db, name);
-            if !member.is_unbound() {
-                return member;
+        for superclass in self.proper_superclasses(db) {
+            match superclass {
+                // TODO we may instead want to record the fact that we encountered dynamic, and intersect it with
+                // the type found on the next "real" class.
+                ClassBase::Any | ClassBase::Unknown | ClassBase::Todo => {
+                    return Type::from(superclass).member(db, name)
+                }
+                ClassBase::Class(class) => {
+                    let member = class.own_class_member(db, name);
+                    if !member.is_unbound() {
+                        return member;
+                    }
+                }
             }
         }
 
         Symbol::Unbound
     }
+
+    /// Iterate through the inferred types of the class's explicit bases.
+    ///
+    /// Note that any class (except for `object`) that has no explicit
+    /// bases will implicitly inherit from `object` at runtime. Nonetheless,
+    /// this method does *not* include `object` in the bases it iterates over.
+    fn bases(self, db: &'db dyn Db) -> impl Iterator<Item = Type<'db>> {
+        let definition = self.definition(db);
+        let class_stmt = self.node(db);
+        let has_type_params = class_stmt.type_params.is_some();
+
+        class_stmt.bases().iter().map(move |base_node| {
+            if has_type_params {
+                let model = SemanticModel::new(db, definition.file(db));
+                base_node.ty(&model)
+            } else {
+                definition_expression_ty(db, definition, base_node)
+            }
+        })
+    }
+
+    /// Return `true` if this class appears to be a cyclic definition,
+    /// i.e., it inherits either directly or indirectly from itself.
+    ///
+    /// A class definition like this will fail at runtime,
+    /// but we must be resilient to it or we could panic.
+    #[salsa::tracked]
+    fn is_cyclically_defined(self, db: &'db dyn Db) -> bool {
+        fn is_cyclically_defined_recursive<'db>(
+            db: &'db dyn Db,
+            class: ClassType<'db>,
+            mut classes_to_watch: FxHashSet<ClassType<'db>>,
+        ) -> bool {
+            if !classes_to_watch.insert(class) {
+                return true;
+            }
+            class
+                .bases(db)
+                .filter_map(Type::into_class_literal_type)
+                .any(|base_class| {
+                    is_cyclically_defined_recursive(db, base_class, classes_to_watch.clone())
```

Comment:
What's the reason for cloning `classes_to_watch` here? Doesn't that mean that we could miss cycles if they're more nested?

Reply:
Without the clone, we start complaining that classes that use diamond inheritance are cyclically defined, because the same class appears in the MROs of two different bases. But that doesn't necessarily mean that the class is cyclic; Python can resolve the MRO for a class that uses diamond inheritance.

> Doesn't that mean that we miss cycles if they're very nested?

I don't _think_ so. Do you have an example of something that you think this logic might miss?

Reply:
Oh interesting. I forgot about diamond inheritance chains.


I think what we could do here instead of cloning is to use an index map and instead always truncate the set when we're done visiting a subgraph to the length it had before visiting that subgraph (which is the same as dropping the entire subgraph)

Reply:
Oh very cool idea! Would never have thought of that. I'll try that out, thanks!

Reply:
Equivalent to that would be for `is_cyclically_defined_recursive` to remove `class` from `classes_to_watch` before returning. But Micha's idea to use an `IndexSet` will be more efficient.